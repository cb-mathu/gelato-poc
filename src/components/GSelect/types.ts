import { VueComponentProps } from '../../types';
import { ExtractPropTypes, SelectHTMLAttributes } from 'vue';
import { SELECT_PROPS } from './constants';

export type SelectProps = ExtractPropTypes<typeof SELECT_PROPS>;
export type NativeSelectProps = Omit<SelectHTMLAttributes, keyof SelectProps>;
export type MergedSelectProps = Partial<SelectProps & VueComponentProps & NativeSelectProps>;
