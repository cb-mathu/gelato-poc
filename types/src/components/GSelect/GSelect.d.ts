import { MergedSelectProps } from './types';
declare const _default: new () => {
    $props: MergedSelectProps;
};
export default _default;
