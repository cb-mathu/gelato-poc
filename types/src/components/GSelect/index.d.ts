import GSelect from './GSelect';
export { GSelect };
