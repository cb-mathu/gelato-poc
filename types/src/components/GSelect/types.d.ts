import { VueComponentProps } from '../../types';
import { ExtractPropTypes, SelectHTMLAttributes } from 'vue';
import { SELECT_PROPS } from './constants';
export declare type SelectProps = ExtractPropTypes<typeof SELECT_PROPS>;
export declare type NativeSelectProps = Omit<SelectHTMLAttributes, keyof SelectProps>;
export declare type MergedSelectProps = Partial<SelectProps & VueComponentProps & NativeSelectProps>;
