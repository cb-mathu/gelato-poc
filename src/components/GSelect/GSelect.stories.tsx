import { ref } from 'vue';
import GSelect from './GSelect';
import { GNotification } from '../GNotification';

export default {
  title: 'Component/Forms/GSelect',
  component: GSelect,
  argTypes: {
    name: {
      default: 'Name of the field',
    },
    label: {
      description: 'Label for the field',
      control: {
        type: 'text',
      },
    },
    description: {
      description: 'Description for the field',
      control: {
        type: 'text',
      },
    },
    disabled: {
      description: 'Disable the field',
      control: {
        type: 'boolean',
      },
    },
    autofocus: {
      description: 'Autofocus the field',
      default: false,
      control: {
        type: 'boolean',
      },
    },
    required: {
      description: 'Field is required',
      default: false,
      control: {
        type: 'boolean',
      },
    },
    readonly: {
      description: 'Field is readonly',
      default: false,
      control: {
        type: 'boolean',
      },
    },
    modelValue: {
      description: 'v-model value',
      control: {
        type: 'text',
      },
    },
    id: {
      description: 'Id of the field',
      control: {
        type: 'text',
      },
    },
    placeholder: {
      description: 'placeholder of the field',
      control: {
        type: 'text',
      },
    },
    class: {
      description: 'Class of the field',
      defaulValue: '',
      control: {
        type: 'text',
      },
    },
    invalid: {
      description: 'Field is invalid',
      defaulValue: false,
      control: {
        type: 'boolean',
      },
    },
    errorMessage: {
      description: 'Error message for invalid field (Applicable only if invalid is set)',
      defaulValue: '',
      control: {
        type: 'text',
      },
    },
  },
};

const options = () => (
  <>
    <option value="">Select country</option>
    <option value="Germany">Germany</option>
    <option value="India">India</option>
  </>
);

const Template = (args: any) => ({
  setup() {
    let input = ref(args.defaultValue || 'India');

    const attributes = {
      disabled: args.disabled,
      readonly: args.readonly,
      invalid: args.invalid,
      errorMessage: args.errorMessage,
    };

    return () => (
      <div>
        <div style="margin-bottom: 20px;">
          <GSelect {...attributes} v-model={input.value} id="input" label={args.label} description={args.description}>
            {{
              label: () => args.labelSlot,
              default: () => args.descriptionSlot,
              options: options(),
            }}
          </GSelect>
        </div>
        <GNotification>
          Selected Value:<b>{input.value}</b>
        </GNotification>
      </div>
    );
  },
});

//Story
export const Story = Template.bind({});

export const DefaultSelect = (props?: any) => (
  <GSelect {...props}>
    {{
      options,
    }}
  </GSelect>
);

export const SelectWithLabel = (props?: any) => (
  <GSelect {...props} label="Select Country">
    {{
      options,
    }}
  </GSelect>
);

export const SelectWithLabelAsSlot = () => (
  <GSelect>
    {{
      label: () => <span>Select Country</span>,
      options,
    }}
  </GSelect>
);

export const SelectWithDescription = () => (
  <GSelect description="Will be used in tax calculation">
    {{
      label: () => <span>Select Country</span>,
      default: () => <span>Select Country</span>,
      options,
    }}
  </GSelect>
);

export const SelectWithDescriptionAsSlot = () => (
  <GSelect>
    {{
      label: () => <span>Select Country</span>,
      default: () => <span>Will be used in tax calculation</span>,
      options,
    }}
  </GSelect>
);

export const DisabledSelect = () => (
  <GSelect disabled>
    {{
      label: () => <span>Select Country</span>,
      default: () => <span>Will be used in tax calculation</span>,
      options,
    }}
  </GSelect>
);

export const ReadonlySelect = () => (
  <GSelect readonly>
    {{
      label: () => <span>Select Country</span>,
      default: () => <span>Will be used in tax calculation</span>,
      options,
    }}
  </GSelect>
);

export const InvalidSelect = () => (
  <GSelect invalid errorMessage="This is an error message, ooh poor you!">
    {{
      label: () => <span>Select Country</span>,
      default: () => <span>Will be used in tax calculation</span>,
      options,
    }}
  </GSelect>
);
