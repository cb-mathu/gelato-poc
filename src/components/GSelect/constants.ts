import { PropType } from 'vue';
import { BASE_FORM_PROPS } from '../../utils/props';

export const SELECT_PROPS = {
  ...BASE_FORM_PROPS,
  label: {
    type: String,
  },
  description: {
    type: String,
  },
  onUpdate: Function as PropType<(event: Event) => void>,
};
