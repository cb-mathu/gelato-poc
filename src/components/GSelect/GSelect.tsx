import { defineComponent } from 'vue';
import { SELECT_PROPS } from './constants';
import { useFormAria, useFormAriaIds } from '../../utils/accessibility';
import { GInputErrorMessage } from '../GInput';
import { useId, useVModel } from '../../composables';
import { MergedSelectProps } from './types';
import { isNotEmpty } from '../../utils/common';

const GSelect = defineComponent({
  name: 'GSelect',
  props: SELECT_PROPS,

  setup(props, { attrs, slots }) {
    const internalValue = useVModel(props, 'modelValue');
    const { labelId, descriptionId, errorId } = useFormAriaIds(props.id);

    return () => {
      const label = props.label || slots.label?.();
      const description = props.description || slots.default?.();

      const propsWeControl = {
        id: props.id || useId('GSelect'),
      };

      const attributes = { ...attrs, ...props, ...propsWeControl };

      const ariaAttributes = {
        ...useFormAria({
          labelledBy: labelId,
          describedBy: descriptionId,
          invalid: props.invalid,
          errorMessage: errorId,
        }),
      };

      const classes = {
        [`g-input`]: true,
        [`g-input--invalid`]: attributes.invalid,
      };
      return (
        <div class={classes}>
          {isNotEmpty(label) && (
            <label class="g-input__label" id={labelId} for={attributes.id}>
              {label}
            </label>
          )}
          {isNotEmpty(description) && (
            <div class="g-input__desc" id={descriptionId}>
              {description}
            </div>
          )}
          <div class="g-input__group">
            <select {...attributes} {...ariaAttributes} v-model={internalValue.value} class="g-input__element">
              {slots?.options?.()}
            </select>
          </div>
          {props.invalid && <GInputErrorMessage id={errorId} errorMessage={props.errorMessage} />}
        </div>
      );
    };
  },
});

export default GSelect as {
  new (): {
    $props: MergedSelectProps;
  };
};
