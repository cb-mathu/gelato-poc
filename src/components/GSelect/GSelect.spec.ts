import { render, screen, fireEvent } from '@testing-library/vue';
import userEvent from '@testing-library/user-event';
import * as stories from './GSelect.stories';

const {
  DefaultSelect,
  SelectWithDescription,
  SelectWithDescriptionAsSlot,
  SelectWithLabel,
  SelectWithLabelAsSlot,
  InvalidSelect,
  ReadonlySelect,
  DisabledSelect,
} = stories;

describe('GSelect : ', () => {
  test('renders the default with default args', () => {
    const { container } = render(DefaultSelect());
    expect(container.firstChild).toMatchSnapshot();
  });

  test('renders with label passed via prop', () => {
    render(SelectWithLabel());
    expect(screen.getByText('Select Country')).toBeVisible();
  });

  test('renders with label passed via slot', () => {
    render(SelectWithLabelAsSlot());
    expect(screen.getByText('Select Country')).toBeVisible();
  });

  test('renders with label and description passed via prop', () => {
    render(SelectWithDescription());
    expect(screen.getByText('Select Country')).toBeVisible();
  });

  test('renders with label and description passed via slot', () => {
    render(SelectWithDescriptionAsSlot());
    expect(screen.getByText('Select Country')).toBeVisible();
  });

  test('renders in disabled state', () => {
    render(DisabledSelect());
    expect(screen.getByRole('combobox')).toHaveAttribute('disabled');
  });

  test('renders in readonly state', () => {
    render(ReadonlySelect());
    expect(screen.getByRole('combobox')).toHaveAttribute('readonly');
  });

  test('renders in invalid state', () => {
    render(InvalidSelect());
    expect(screen.getByText('This is an error message, ooh poor you!')).toBeVisible();
  });

  test('pressing tab will focus select', () => {
    render(SelectWithLabel());
    userEvent.tab();
    expect(screen.getByRole('combobox')).toHaveFocus();
  });

  test('focuses select when label is clicked', () => {
    render(SelectWithLabel());
    const element = screen.getByText('Select Country');
    userEvent.click(element);
    expect(screen.getByRole('combobox')).toHaveFocus();
  });

  test('does not focuses disabled select when label is clicked', () => {
    render(SelectWithLabel({ disabled: true }));
    const element = screen.getByText('Select Country');
    userEvent.click(element);
    expect(screen.getByRole('combobox')).not.toHaveFocus();
  });

  test('updates the state on option select', async () => {
    render(DefaultSelect({ modelValue: 'Germany' }));
    const element = screen.getByRole('combobox');
    userEvent.selectOptions(element, ['India']);
    expect(element).toHaveValue('India');
  });
});
