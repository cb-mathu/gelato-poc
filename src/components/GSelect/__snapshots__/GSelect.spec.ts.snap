// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GSelect :  renders the default with default args 1`] = `
<div
  class="g-input"
>
  <!---->
  <!---->
  <div
    class="g-input__group"
  >
    <select
      aria-describedby="description_2"
      aria-errormessage=""
      aria-invalid="false"
      aria-labelledby="label_1"
      autocomplete="off"
      class="g-input__element"
      id="GSelect_4"
      invalid="false"
      modelmodifiers="[object Object]"
      modelvalue=""
      name=""
      placeholder=""
    >
      
      
      <option
        value=""
      >
        Select country
      </option>
      <option>
        Germany
      </option>
      <option>
        India
      </option>
      
      
    </select>
  </div>
  <!---->
</div>
`;
